### PR TITLE
 #4911 - insert image from ket to canvas

### PR DESCRIPTION
**Pull request by daniil-sloboda**
* Loading image from ket file to canvas
* Saving image from struct to ket
* operations to move image during paste\replace

## How the feature works? / How did you fix the issue?
(Screenshots, videos, or GIFs, if applicable)

## Check list
- [ ] unit-tests written
- [ ] e2e-tests written
- [ ] documentation updated
- [x] PR name follows the pattern `#1234 – issue name`
- [x] branch name doesn't contain '#'
- [x] PR is linked with the issue
- [x] base branch (master or release/xx) is correct
- [x] task status changed to "Code review"
- [x] reviewers are notified about the pull request

### CODE REVIEW

File: packages/ketcher-core/src/application/editor/operations/rasterImage/rasterImageUpsertDelete.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    super(OperationType.RASTER_IMAGE_UPSERT);
+  }
+
+  execute(reStruct: ReStruct) {
+    const struct = reStruct.molecule;
+
+    if (!this.id) {
+      this.id = struct.rasterImages.newId();
+    }
+    const id = this.id!;
```
Comment (rrodionov91) [inline packages/ketcher-core/src/application/editor/operations/rasterImage/rasterImageUpsertDelete.ts:35]: I believe "!" is excessive here

File: packages/ketcher-core/src/application/editor/operations/rasterImage/rasterImageUpsertDelete.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    const item = this.rasterImage.clone();
+    struct.rasterImages.set(id, item);
+    reStruct.rasterImages.set(id, new ReRasterImage(item));
+
+    BaseOperation.invalidateItem(reStruct, 'rasterImages', id, 1);
```
Comment (rrodionov91) [inline packages/ketcher-core/src/application/editor/operations/rasterImage/rasterImageUpsertDelete.ts:40]: Please move 'rasterImages' to constant

Comment (rrodionov91) [reply packages/ketcher-core/src/application/editor/operations/rasterImage/rasterImageUpsertDelete.ts:40]: Of better to object with constants to use it for all such invalidation items names in future (including existing ones)

File: packages/ketcher-core/src/application/render/restruct/rerasterImage.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+export class ReRasterImage extends ReObject {
+  private element?: RaphaelElement;
+
+  static isSelectable(): boolean {
+    return true;
+  }
+
+  constructor(private item: RasterImage) {
+    super('image');
```
Comment (rrodionov91) [inline packages/ketcher-core/src/application/render/restruct/rerasterImage.ts:16]: Please create constant or object with constants to reuse for all such scenarios in future

File: packages/ketcher-core/src/application/render/restruct/rerasterImage.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  }
+
+  show(restruct: ReStruct, options: RenderOptions) {
+    if (this.element) {
+      return;
+    }
```
Comment (rrodionov91) [inline packages/ketcher-core/src/application/render/restruct/rerasterImage.ts:22]: Probably it is more safely to not return in this case, but rerender? Just let's consider.

File: packages/ketcher-core/src/domain/entities/rasterImage.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+  clone(): RasterImage {
+    return new RasterImage(this.bitmap, this.position);
+  }
+
+  addPositionOffset(offset: Vec2) {
+    this.position = this.position.map((item) => item.add(offset)) as Position;
+  }
+
+  scaled(scale: number): void {
```
Comment (rrodionov91) [inline packages/ketcher-core/src/domain/entities/rasterImage.ts:35]: I would suggest rename it to something like "rescalePosition"

File: packages/ketcher-core/src/domain/serializers/ket/schema.json [hunk 6]
```diff
@@ -777,6 +780,41 @@
...
+                "properties": {
+                  "x": {
+                    "type": "number"
+                  },
+                  "y": {
+                    "type": "number"
+                  },
+                  "z": {
+                    "type": "number"
+                  }
```
Comment (rrodionov91) [inline packages/ketcher-core/src/domain/serializers/ket/schema.json:811]: Do we need "z"?

File: packages/ketcher-core/src/domain/serializers/ket/toKet/prepare.ts [hunk 7]
```diff
@@ -96,6 +96,19 @@ export function prepareStructForKet(struct: Struct) {
     });
   });
 
+  struct.rasterImages.forEach((image) => {
+    const center: Vec2 = Vec2.centre(image.position[0], image.position[1]);
+    ketNodes.push({
+      type: 'rasterImage',
```
Comment (rrodionov91) [inline packages/ketcher-core/src/domain/serializers/ket/toKet/prepare.ts:102]: Please use constant

File: packages/ketcher-core/src/domain/serializers/ket/toKet/rasterImageToKet.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ***************************************************************************/
+
+import { getNodeWithInvertedYCoord } from '../helpers';
+
+export function rasterImageToKet(rasterImageNode) {
+  return {
+    type: 'rasterImage',
```
Comment (rrodionov91) [inline packages/ketcher-core/src/domain/serializers/ket/toKet/rasterImageToKet.ts:21]: Please use constant

**Review by rrodionov91 — CHANGES_REQUESTED** [review]

File: packages/ketcher-core/src/application/editor/operations/rasterImage/rasterImageUpsertDelete.ts (hunk 1 above)
Comment (daniil-sloboda) [reply packages/ketcher-core/src/application/editor/operations/rasterImage/rasterImageUpsertDelete.ts:35]: Sadly current typescript version used in the project cannot figure out that id cannot be undefined on the line 35

File: packages/ketcher-core/src/application/editor/operations/rasterImage/rasterImageMove.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    if (!item || !renderItem) {
+      return;
+    }
+    const differenceScaled = Scale.modelToCanvas(
+      this.difference,
+      reStruct.render.options,
+    );
+
+    renderItem.move(differenceScaled);
+    item.addPositionOffset(this.difference);
```
[the remark this answers is not in the document]
Comment (daniil-sloboda) [reply packages/ketcher-core/src/application/editor/operations/rasterImage/rasterImageMove.ts:25]: raphael sadly have no method to set exact position
https://dmitrybaranovskiy.github.io/raphael/reference.html#Element

**Review by rrodionov91 — APPROVED** [review]

File: packages/ketcher-core/src/domain/entities/rasterImage.ts [hunk 10]
```diff
@@ -19,20 +19,25 @@ import { Vec2 } from 'domain/entities/vec2';
...
   constructor(public bitmap: string, public position: [Vec2, Vec2]) {
     super();
   }
 
   clone(): RasterImage {
-    return new RasterImage(this.bitmap, this.position);
+    return new RasterImage(
+      this.bitmap,
+      this.position.map((item) => item) as Position,
+    );
```
Comment (rrodionov91) [inline packages/ketcher-core/src/domain/entities/rasterImage.ts:33]: Looks like position.item still sharing between instance and it's clone

**Review by rrodionov91 — APPROVED** [review]

**Review by rrodionov91 — APPROVED** [review]

**Pull request merged**